File: iOS/dimina/DiminaKit/Container/Api/Network/WebSocketAPI.swift
//
//  WebSocketAPI.swift
//  dimina
//
//  Bridge shim for wx.connectSocket / SocketTask.
//
//  This file only wires the 11 bridge method names to
//  `DMPWebSocketManager.shared`; all behavior (validation, state machine,
//  background policy, legacy binding, ...) lives in DMPWebSocketManager /
//  DMPWebSocketValidation.

import Foundation

public class WebSocketAPI: DMPContainerApi {

    private static let CONNECT_SOCKET = "connectSocket"
    private static let SEND_SOCKET_MESSAGE = "sendSocketMessage"
    private static let CLOSE_SOCKET = "closeSocket"
    private static let ON_SOCKET_OPEN = "onSocketOpen"
    private static let ON_SOCKET_MESSAGE = "onSocketMessage"
    private static let ON_SOCKET_ERROR = "onSocketError"
    private static let ON_SOCKET_CLOSE = "onSocketClose"
    private static let OFF_SOCKET_OPEN = "offSocketOpen"
    private static let OFF_SOCKET_MESSAGE = "offSocketMessage"
    private static let OFF_SOCKET_ERROR = "offSocketError"
    private static let OFF_SOCKET_CLOSE = "offSocketClose"

    @BridgeMethod(CONNECT_SOCKET)
    var connectSocket: DMPBridgeMethodHandler = { param, env, callback in
        let app = DMPAppManager.sharedInstance().getApp(appIndex: env.appIndex)
        // app.json 的 `networkTimeout.connectSocket` 是这个小程序自己的连接超时默认值，
        // 调用方传的 timeout 优先于它，两者都没有时才是 60000。
        DMPWebSocketManager.shared.connectSocket(
            params: param.getMap(),
            appId: env.appId,
            appVersion: app?.jsAppVersion() ?? "0",
            appNetworkTimeoutMs: app?.getBundleAppConfig()?.networkTimeoutConnectSocketMs,
            callback: callback
        )
        return DMPNoneResult()
    }

    @BridgeMethod(SEND_SOCKET_MESSAGE)
    var sendSocketMessage: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.sendSocketMessage(params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(CLOSE_SOCKET)
    var closeSocket: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.closeSocket(params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(ON_SOCKET_OPEN)
    var onSocketOpen: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.onSocketEvent(event: .open, params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(ON_SOCKET_MESSAGE)
    var onSocketMessage: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.onSocketEvent(event: .message, params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(ON_SOCKET_ERROR)
    var onSocketError: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.onSocketEvent(event: .error, params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(ON_SOCKET_CLOSE)
    var onSocketClose: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.onSocketEvent(event: .close, params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(OFF_SOCKET_OPEN)
    var offSocketOpen: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.offSocketEvent(event: .open, params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(OFF_SOCKET_MESSAGE)
    var offSocketMessage: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.offSocketEvent(event: .message, params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(OFF_SOCKET_ERROR)
    var offSocketError: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.offSocketEvent(event: .error, params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }

    @BridgeMethod(OFF_SOCKET_CLOSE)
    var offSocketClose: DMPBridgeMethodHandler = { param, env, callback in
        DMPWebSocketManager.shared.offSocketEvent(event: .close, params: param.getMap(), appId: env.appId, callback: callback)
        return DMPNoneResult()
    }
}

File: android/dimina/src/main/kotlin/com/didi/dimina/api/network/WebSocketApi.kt
package com.didi.dimina.api.network

import com.didi.dimina.api.APIResult
import com.didi.dimina.api.BaseApiHandler
import com.didi.dimina.api.NoneResult
import com.didi.dimina.ui.container.DiminaActivity
import org.json.JSONObject

/**
 * Bridge shell for `wx.connectSocket`/`SocketTask`.
 *
 * Pure wiring: figures out which owner (appId) is calling, whether this is a task-mode call
 * (has `socketId`) or a legacy global-mode call, refreshes the event push channel, and delegates
 * everything else to [WebSocketManager]. All 11 API names always return [NoneResult] - one-shot
 * success/fail/complete callbacks are triggered by the Manager itself via `ApiUtils`, matching
 * `NetworkApi`'s `REQUEST` path, and avoiding `MiniApp.invokeAPI`'s `AsyncResult` auto-dispatch.
 */
class WebSocketApi : BaseApiHandler() {
    companion object {
        private const val CONNECT_SOCKET = "connectSocket"
        private const val SEND_SOCKET_MESSAGE = "sendSocketMessage"
        private const val CLOSE_SOCKET = "closeSocket"
        private const val ON_SOCKET_OPEN = "onSocketOpen"
        private const val ON_SOCKET_MESSAGE = "onSocketMessage"
        private const val ON_SOCKET_ERROR = "onSocketError"
        private const val ON_SOCKET_CLOSE = "onSocketClose"
        private const val OFF_SOCKET_OPEN = "offSocketOpen"
        private const val OFF_SOCKET_MESSAGE = "offSocketMessage"
        private const val OFF_SOCKET_ERROR = "offSocketError"
        private const val OFF_SOCKET_CLOSE = "offSocketClose"
    }

    override val apiNames = setOf(
        CONNECT_SOCKET, SEND_SOCKET_MESSAGE, CLOSE_SOCKET,
        ON_SOCKET_OPEN, ON_SOCKET_MESSAGE, ON_SOCKET_ERROR, ON_SOCKET_CLOSE,
        OFF_SOCKET_OPEN, OFF_SOCKET_MESSAGE, OFF_SOCKET_ERROR, OFF_SOCKET_CLOSE,
    )

    override fun handleAction(
        activity: DiminaActivity,
        appId: String,
        apiName: String,
        params: JSONObject,
        responseCallback: (String) -> Unit,
    ): APIResult {
        val manager = WebSocketManager.shared
        // Every invocation refreshes the emitter closure so persistent events can still reach
        // this appId's service JsCore even after the invoking page/Bridge is gone.
        manager.updateEmitter(appId) { json -> responseCallback(json) }

        // The same API name carries both task-mode (has socketId) and legacy global-mode (no
        // socketId) calls; branch on params, not on a second API name.
        val hasSocketId = params.has("socketId")

        when (apiName) {
            CONNECT_SOCKET -> manager.connectSocket(
                appId,
                activity.getMiniProgram().versionCode.toString(),
                params,
                responseCallback,
            )
            SEND_SOCKET_MESSAGE -> manager.sendSocketMessage(appId, hasSocketId, params, responseCallback)
            CLOSE_SOCKET -> manager.closeSocket(appId, hasSocketId, params, responseCallback)
            ON_SOCKET_OPEN -> manager.onSocketEvent("open", appId, hasSocketId, params, apiName, responseCallback)
            ON_SOCKET_MESSAGE -> manager.onSocketEvent("message", appId, hasSocketId, params, apiName, responseCallback)
            ON_SOCKET_ERROR -> manager.onSocketEvent("error", appId, hasSocketId, params, apiName, responseCallback)
            ON_SOCKET_CLOSE -> manager.onSocketEvent("close", appId, hasSocketId, params, apiName, responseCallback)
            OFF_SOCKET_OPEN -> manager.offSocketEvent("open", appId, hasSocketId, params, apiName, responseCallback)
            OFF_SOCKET_MESSAGE -> manager.offSocketEvent("message", appId, hasSocketId, params, apiName, responseCallback)
            OFF_SOCKET_ERROR -> manager.offSocketEvent("error", appId, hasSocketId, params, apiName, responseCallback)
            OFF_SOCKET_CLOSE -> manager.offSocketEvent("close", appId, hasSocketId, params, apiName, responseCallback)
        }

        return NoneResult()
    }
}
